client.go

```diff
@@ -120,6 +147,79 @@ func (c *Client) Push(n *Notification) (*Response, error) {
 	return response, nil
 }
 
+// EnablePinging tries to send PING frames to APNs servers whenever the client
+// has a valid connection. If the willHandleDrops parameter is set to true, this
+// function returns a read-only channel that gets notified when pinging fails.
+// This allows the user to take actions to preemptively reinitialize the client's
+// connection. The second return value indicates whether the call has successfully
+// enabled pinging.
+func (c *Client) EnablePinging(willHandleDrops bool) (<-chan struct{}, bool) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	if c.pinging {
+		return nil, false
+	}
+	c.pinging = true
+	var dropSignal chan struct{}
+	if willHandleDrops {
+		dropSignal = make(chan struct{})
+	}
+	go func() {
+		// 8 bytes of random data used for PING-PONG, as per HTTP/2 spec.
+		data := [8]byte{byte(rand.Intn(256)), byte(rand.Intn(256)), byte(rand.Intn(256)), byte(rand.Intn(256)), byte(rand.Intn(256)), byte(rand.Intn(256)), byte(rand.Intn(256)), byte(rand.Intn(256))}
+		pinger := new(time.Ticker)
+		var framer *http2.Framer
+		c.m.Lock()
+		if c.conn != nil {
+			framer = http2.NewFramer(c.conn, c.conn)
+			pinger = time.NewTicker(PingPongFrequency)
+		}
+		c.m.Unlock()
+		for {
+			select {
+			case <-pinger.C:
+				err := framer.WritePing(false, data)
+				if err != nil {
+					// APNs did not answer with pong, which means the connection
+					// has been dropped. Stop trying and notify the drop handler,
+					// if there is any.
+					c.m.Lock()
+					c.conn = nil
+					c.m.Unlock()
+					pinger.Stop()
+					if willHandleDrops {
+						dropSignal <- struct{}{}
+					}
+				}
+			case <-c.newConnChan:
+				c.m.Lock()
+				framer = http2.NewFramer(c.conn, c.conn)
```

Comment:
[The source code](https://github.com/sideshow/apns2/issues/20#issuecomment-216501289) and many production tests reveal that it does not.